packages/shared/scope-hoisting/src/formats/esmodule.js

```diff
@@ -325,9 +324,85 @@ export function generateExports(
           }
         }
       }
+      exportedSymbols.forEach(s => exportedIdentifiers.delete(s));
     },
   });
 
+  if (exportedIdentifiers.size > 0) {
+    let declarations = [];
+    let exportedIdentifiersSpecifiers = [];
+    // `export { $id$init().foo as foo};` is not valid, so instead do:
+    // ```
+    // let syntheticExport$foo = $id$init().foo;
+    // export { syntheticExport$foo as foo};
+    // ```
+    for (let [exportAs, symbol] of exportedIdentifiers) {
+      declarations.push(
+        t.variableDeclarator(
+          t.identifier('syntheticExport$' + exportAs),
+          t.identifier(symbol),
+        ),
+      );
+      exportedIdentifiersSpecifiers.push(
+        t.exportSpecifier(
+          t.identifier('syntheticExport$' + exportAs),
+          t.identifier(exportAs),
+        ),
+      );
+    }
+    let [decl, exports] = programPath.pushContainer('body', [
+      t.variableDeclaration('var', declarations),
+      t.exportNamedDeclaration(null, exportedIdentifiersSpecifiers),
+    ]);
+    invariant(isVariableDeclaration(decl.node));
+    programPath.scope.registerDeclaration(decl);
+    for (let d of decl.get<Array<NodePath<VariableDeclarator>>>(
+      'declarations',
+    )) {
+      maybeReplaceIdentifier(d.get<NodePath<Identifier>>('init'));
+    }
+    invariant(isExportNamedDeclaration(exports.node));
+    programPath.scope.registerDeclaration(exports);
+    for (let e of exports.get<Array<NodePath<ExportSpecifier>>>('specifiers')) {
+      nullthrows(programPath.scope.getBinding(e.node.local.name)).reference(
+        e.get<NodePath<Identifier>>('local'),
+      );
+    }
+  }
+
+  if (exportedIdentifiersBailout.size > 0) {
```

Comment:
A esmodule bundle could reexport something from an asset in different bundle, but I don't think this can actually happen at the moment (and there are apparently no tests for this). Should I just remove that branch with an assertion for now?

Reply:
Either that or add a comment?

Reply:
I've commented out the code and added back the error/assertion that was thrown previously in this situation.